File: python/ccxt/async_support/base/multi_throttler.py
import asyncio
import time
from typing import Dict, List, Optional, Any


class ThrottleRule:
    """Represents a single rate limiting rule"""
    
    def __init__(self, id: str, capacity: float, refill_rate: float, tokens: float, 
                 interval_type: str, interval_num: int):
        self.id = id  # Unique identifier for the rule
        self.capacity = capacity  # Maximum tokens this rule can hold
        self.refill_rate = refill_rate  # Rate at which tokens are refilled (tokens per millisecond)
        self.tokens = tokens  # Current available tokens
        self.interval_type = interval_type  # For documentation/debugging ('SECOND', 'MINUTE', 'HOUR', 'DAY')
        self.interval_num = interval_num  # Number of intervals


class MultiThrottlerConfig:
    """Configuration options for the MultiThrottler"""
    
    def __init__(self, max_capacity: int = 2000, delay: float = 0.001):
        self.max_capacity = max_capacity  # Maximum queue size before throwing errors
        self.delay = delay  # Sleep delay between checks (in seconds)


class QueueItem:
    """Queue item representing a pending request"""
    
    def __init__(self, resolver, cost: Dict[str, float], timestamp: float):
        self.resolver = resolver
        self.cost = cost
        self.timestamp = timestamp


class MultiThrottler:
    """
    Multi-rule throttler that can enforce multiple rate limiting rules simultaneously.
    Supports Binance-style rate limiting with different rule types (REQUEST_WEIGHT, RAW_REQUESTS, ORDERS, etc.)
    """
    
    def __init__(self, rules: List[ThrottleRule], config: Optional[MultiThrottlerConfig] = None):
        self.rules: Dict[str, ThrottleRule] = {}
        
        # Initialize rules map
        for rule in rules:
            # Clone rule to avoid mutations
            cloned_rule = ThrottleRule(
                rule.id, rule.capacity, rule.refill_rate, 
                rule.tokens, rule.interval_type, rule.interval_num
            )
            self.rules[rule.id] = cloned_rule

        self.config = config or MultiThrottlerConfig()
        self.queue: List[QueueItem] = []
        self.running = False

    def add_rule(self, rule: ThrottleRule) -> None:
        """Add or update a throttling rule"""
        cloned_rule = ThrottleRule(
            rule.id, rule.capacity, rule.refill_rate, 
            rule.tokens, rule.interval_type, rule.interval_num
        )
        self.rules[rule.id] = cloned_rule

    def remove_rule(self, rule_id: str) -> bool:
        """Remove a throttling rule"""
        return self.rules.pop(rule_id, None) is not None

    def get_status(self) -> Dict[str, Dict[str, float]]:
        """Get current status of all rules"""
        status = {}
        
        for rule_id, rule in self.rules.items():
            status[rule_id] = {
                'tokens': rule.tokens,
                'capacity': rule.capacity,
                'utilization': 1 - (rule.tokens / rule.capacity)
            }
        
        return status

    def _can_process(self, cost: Dict[str, float]) -> bool:
        """Check if a request can be processed immediately (all rules have sufficient tokens)"""
        for rule_id, rule_cost in cost.items():
            rule = self.rules.get(rule_id)
            if not rule:
                raise ValueError(f"Unknown throttle rule: {rule_id}")
            
            if rule.tokens < rule_cost:
                return False
        return True

    def _consume_tokens(self, cost: Dict[str, float]) -> None:
        """Consume tokens from all applicable rules"""
        for rule_id, rule_cost in cost.items():
            rule = self.rules.get(rule_id)
            if rule:
                rule.tokens -= rule_cost

    def _refill_tokens(self, elapsed: float) -> None:
        """Refill tokens for all rules based on elapsed time"""
        for rule in self.rules.values():
            tokens_to_add = rule.refill_rate * elapsed
            rule.tokens = min(rule.tokens + tokens_to_add, rule.capacity)

    def _calculate_wait_time(self, cost: Dict[str, float]) -> float:
        """Calculate the minimum time needed for a request to be processable"""
        max_wait_time = 0.0

        for rule_id, rule_cost in cost.items():
            rule = self.rules.get(rule_id)
            if not rule:
                continue

            if rule.tokens < rule_cost:
                tokens_needed = rule_cost - rule.tokens
                wait_time = tokens_needed / rule.refill_rate
                max_wait_time = max(max_wait_time, wait_time)

        return max_wait_time

    async def _loop(self) -> None:
        """Main processing loop"""
        last_timestamp = time.time() * 1000  # Convert to milliseconds

        while self.running and len(self.queue) > 0:
            current_time = time.time() * 1000  # Convert to milliseconds
            elapsed = current_time - last_timestamp
            last_timestamp = current_time

            # Refill tokens for all rules
            self._refill_tokens(elapsed)

            # Process as many items from the queue as possible
            processed = 0
            while len(self.queue) > 0:
                item = self.queue[0]
                
                if self._can_process(item.cost):
                    self._consume_tokens(item.cost)
                    if not item.resolver.done():
                        item.resolver.set_result(None)
                    self.queue.pop(0)
                    processed += 1
                    
                    # Allow other async operations to run
                    if processed % 10 == 0:
                        await asyncio.sleep(0)
                else:
                    # Can't process this item yet, break and wait
                    break

            # If no items were processed, wait before trying again
            if processed == 0 and len(self.queue) > 0:
                wait_time = self._calculate_wait_time(self.queue[0].cost)
                sleep_time = min(wait_time / 1000, self.config.delay)  # Convert to seconds
                await asyncio.sleep(sleep_time)

            # Stop if queue is empty
            if len(self.queue) == 0:
                self.running = False

    async def throttle(self, cost: Dict[str, float]) -> None:
        """
        Submit a request to be throttled according to the defined rules
        
        Args:
            cost: Dictionary mapping rule IDs to their costs for this request
        
        Returns:
            Promise that resolves when the request can proceed
        """
        # Validate that all cost rules exist
        for rule_id in cost.keys():
            if rule_id not in self.rules:
                available_rules = ', '.join(self.rules.keys())
                raise ValueError(f"Unknown throttle rule: {rule_id}. Available rules: {available_rules}")

        # Check queue capacity
        if len(self.queue) >= self.config.max_capacity:
            raise RuntimeError(f'Throttle queue is over maxCapacity ({self.config.max_capacity}), see https://github.com/ccxt/ccxt/issues/11645#issuecomment-1195695526')

        # Create future for this request
        future = asyncio.Future()

        # Add to queue
        item = QueueItem(
            resolver=future,
            cost=cost,
            timestamp=time.time() * 1000
        )
        self.queue.append(item)

        # Start processing loop if not already running
        if not self.running:
            self.running = True
            asyncio.ensure_future(self._loop())  # Don't await here to allow immediate return

        return await future

    def get_queue_length(self) -> int:
        """Get the current queue length"""
        return len(self.queue)

    def is_running(self) -> bool:
        """Check if the throttler is currently running"""
        return self.running

    def reset(self) -> None:
        """Reset all token buckets to their capacity"""
        for rule in self.rules.values():
            rule.tokens = rule.capacity

    def set_tokens(self, rule_id: str, tokens: float) -> None:
        """Manually set tokens for a specific rule (useful for testing)"""
        rule = self.rules.get(rule_id)
        if rule:
            rule.tokens = max(0, min(tokens, rule.capacity))

File: cs/ccxt/base/MultiThrottler.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Linq;

namespace ccxt
{
    /// <summary>
    /// Represents a single rate limiting rule
    /// </summary>
    public class ThrottleRule
    {
        public string Id { get; set; }                     // Unique identifier for the rule
        public double Capacity { get; set; }               // Maximum tokens this rule can hold
        public double RefillRate { get; set; }             // Rate at which tokens are refilled (tokens per millisecond)
        public double Tokens { get; set; }                 // Current available tokens
        public string IntervalType { get; set; }           // For documentation/debugging ('SECOND', 'MINUTE', 'HOUR', 'DAY')
        public int IntervalNum { get; set; }               // Number of intervals

        public ThrottleRule(string id, double capacity, double refillRate, double tokens, string intervalType, int intervalNum)
        {
            Id = id;
            Capacity = capacity;
            RefillRate = refillRate;
            Tokens = tokens;
            IntervalType = intervalType;
            IntervalNum = intervalNum;
        }

        // Clone constructor
        public ThrottleRule(ThrottleRule other)
        {
            Id = other.Id;
            Capacity = other.Capacity;
            RefillRate = other.RefillRate;
            Tokens = other.Tokens;
            IntervalType = other.IntervalType;
            IntervalNum = other.IntervalNum;
        }
    }

    /// <summary>
    /// Configuration options for the MultiThrottler
    /// </summary>
    public class MultiThrottlerConfig
    {
        public int MaxCapacity { get; set; } = 2000;       // Maximum queue size before throwing errors
        public double Delay { get; set; } = 0.001;         // Sleep delay between checks (in seconds)

        public MultiThrottlerConfig() { }

        public MultiThrottlerConfig(int maxCapacity, double delay)
        {
            MaxCapacity = maxCapacity;
            Delay = delay;
        }
    }

    /// <summary>
    /// Queue item representing a pending request
    /// </summary>
    internal class QueueItem
    {
        public TaskCompletionSource<object> TaskCompletionSource { get; set; }
        public Dictionary<string, double> Cost { get; set; }
        public long Timestamp { get; set; }

        public QueueItem(TaskCompletionSource<object> taskCompletionSource, Dictionary<string, double> cost, long timestamp)
        {
            TaskCompletionSource = taskCompletionSource;
            Cost = cost;
            Timestamp = timestamp;
        }
    }

    /// <summary>
    /// Multi-rule throttler that can enforce multiple rate limiting rules simultaneously.
    /// Supports Binance-style rate limiting with different rule types (REQUEST_WEIGHT, RAW_REQUESTS, ORDERS, etc.)
    /// </summary>
    public class MultiThrottler
    {
        private readonly Dictionary<string, ThrottleRule> _rules;
        private readonly MultiThrottlerConfig _config;
        private readonly Queue<QueueItem> _queue;
        private bool _running;
        private readonly object _lock = new object();

        public MultiThrottler(List<ThrottleRule> rules, MultiThrottlerConfig config = null)
        {
            _rules = new Dictionary<string, ThrottleRule>();
            
            // Initialize rules map
            foreach (var rule in rules)
            {
                // Clone rule to avoid mutations
                _rules[rule.Id] = new ThrottleRule(rule);
            }

            _config = config ?? new MultiThrottlerConfig();
            _queue = new Queue<QueueItem>();
            _running = false;
        }

        /// <summary>
        /// Add or update a throttling rule
        /// </summary>
        public void AddRule(ThrottleRule rule)
        {
            lock (_lock)
            {
                _rules[rule.Id] = new ThrottleRule(rule);
            }
        }

        /// <summary>
        /// Remove a throttling rule
        /// </summary>
        public bool RemoveRule(string ruleId)
        {
            lock (_lock)
            {
                return _rules.Remove(ruleId);
            }
        }

        /// <summary>
        /// Get current status of all rules
        /// </summary>
        public Dictionary<string, Dictionary<string, double>> GetStatus()
        {
            var status = new Dictionary<string, Dictionary<string, double>>();
            
            lock (_lock)
            {
                foreach (var kvp in _rules)
                {
                    var rule = kvp.Value;
                    status[kvp.Key] = new Dictionary<string, double>
                    {
                        ["tokens"] = rule.Tokens,
                        ["capacity"] = rule.Capacity,
                        ["utilization"] = 1 - (rule.Tokens / rule.Capacity)
                    };
                }
            }
            
            return status;
        }

        /// <summary>
        /// Check if a request can be processed immediately (all rules have sufficient tokens)
        /// </summary>
        private bool CanProcess(Dictionary<string, double> cost)
        {
            foreach (var kvp in cost)
            {
                if (!_rules.TryGetValue(kvp.Key, out var rule))
                {
                    throw new ArgumentException($"Unknown throttle rule: {kvp.Key}");
                }
                
                if (rule.Tokens < kvp.Value)
                {
                    return false;
                }
            }
            return true;
        }

        /// <summary>
        /// Consume tokens from all applicable rules
        /// </summary>
        private void ConsumeTokens(Dictionary<string, double> cost)
        {
            foreach (var kvp in cost)
            {
                if (_rules.TryGetValue(kvp.Key, out var rule))
                {
                    rule.Tokens -= kvp.Value;
                }
            }
        }

        /// <summary>
        /// Refill tokens for all rules based on elapsed time
        /// </summary>
        private void RefillTokens(double elapsed)
        {
            foreach (var rule in _rules.Values)
            {
                var tokensToAdd = rule.RefillRate * elapsed;
                rule.Tokens = Math.Min(rule.Tokens + tokensToAdd, rule.Capacity);
            }
        }

        /// <summary>
        /// Calculate the minimum time needed for a request to be processable
        /// </summary>
        private double CalculateWaitTime(Dictionary<string, double> cost)
        {
            double maxWaitTime = 0.0;

            foreach (var kvp in cost)
            {
                if (!_rules.TryGetValue(kvp.Key, out var rule))
                    continue;

                if (rule.Tokens < kvp.Value)
                {
                    var tokensNeeded = kvp.Value - rule.Tokens;
                    var waitTime = tokensNeeded / rule.RefillRate;
                    maxWaitTime = Math.Max(maxWaitTime, waitTime);
                }
            }

            return maxWaitTime;
        }

        /// <summary>
        /// Main processing loop
        /// </summary>
        private async Task Loop()
        {
            long lastTimestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

            while (_running)
            {
                QueueItem item = null;
                
                lock (_lock)
                {
                    if (_queue.Count == 0)
                    {
                        _running = false;
                        return;
                    }
                }

                long currentTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
                double elapsed = currentTime - lastTimestamp;
                lastTimestamp = currentTime;

                lock (_lock)
                {
                    // Refill tokens for all rules
                    RefillTokens(elapsed);

                    // Process as many items from the queue as possible
                    int processed = 0;
                    while (_queue.Count > 0)
                    {
                        item = _queue.Peek();
                        
                        if (CanProcess(item.Cost))
                        {
                            ConsumeTokens(item.Cost);
                            _queue.Dequeue();
                            
                            // Complete the task outside the lock
                            Task.Run(() => item.TaskCompletionSource.SetResult(null));
                            
                            processed++;
                            
                            // Allow other operations to run
                            if (processed % 10 == 0)
                            {
                                break;
                            }
                        }
                        else
                        {
                            // Can't process this item yet, break and wait
                            break;
                        }
                    }

                    // If no items were processed, calculate wait time
                    if (processed == 0 && _queue.Count > 0)
                    {
                        item = _queue.Peek();
                    }
                    else
                    {
                        item = null;
                    }
                }

                if (item != null)
                {
                    var waitTime = CalculateWaitTime(item.Cost);
                    var sleepTime = Math.Min(waitTime, _config.Delay * 1000);
                    await Task.Delay(TimeSpan.FromMilliseconds(sleepTime));
                }
                else
                {
                    // Small delay to prevent tight loop
                    await Task.Delay(1);
                }
            }
        }

        /// <summary>
        /// Submit a request to be throttled according to the defined rules
        /// </summary>
        /// <param name="cost">Dictionary mapping rule IDs to their costs for this request</param>
        /// <returns>Task that completes when the request can proceed</returns>
        public async Task Throttle(Dictionary<string, double> cost)
        {
            // Validate that all cost rules exist
            foreach (var ruleId in cost.Keys)
            {
                if (!_rules.ContainsKey(ruleId))
                {
                    var availableRules = string.Join(", ", _rules.Keys);
                    throw new ArgumentException($"Unknown throttle rule: {ruleId}. Available rules: {availableRules}");
                }
            }

            TaskCompletionSource<object> taskCompletionSource;
            
            lock (_lock)
            {
                // Check queue capacity
                if (_queue.Count >= _config.MaxCapacity)
                {
                    throw new InvalidOperationException($"Throttle queue is over maxCapacity ({_config.MaxCapacity}), see https://github.com/ccxt/ccxt/issues/11645#issuecomment-1195695526");
                }

                // Create task completion source for this request
                taskCompletionSource = new TaskCompletionSource<object>();

                // Add to queue
                var item = new QueueItem(
                    taskCompletionSource,
                    cost,
                    DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
                );
                _queue.Enqueue(item);

                // Start processing loop if not already running
                if (!_running)
                {
                    _running = true;
                    Task.Run(Loop); // Don't await here to allow immediate return
                }
            }

            await taskCompletionSource.Task;
        }

        /// <summary>
        /// Get the current queue length
        /// </summary>
        public int GetQueueLength()
        {
            lock (_lock)
            {
                return _queue.Count;
            }
        }

        /// <summary>
        /// Check if the throttler is currently running
        /// </summary>
        public bool IsRunning()
        {
            lock (_lock)
            {
                return _running;
            }
        }

        /// <summary>
        /// Reset all token buckets to their capacity
        /// </summary>
        public void Reset()
        {
            lock (_lock)
            {
                foreach (var rule in _rules.Values)
                {
                    rule.Tokens = rule.Capacity;
                }
            }
        }

        /// <summary>
        /// Manually set tokens for a specific rule (useful for testing)
        /// </summary>
        public void SetTokens(string ruleId, double tokens)
        {
            lock (_lock)
            {
                if (_rules.TryGetValue(ruleId, out var rule))
                {
                    rule.Tokens = Math.Max(0, Math.Min(tokens, rule.Capacity));
                }
            }
        }
    }
}